pkg/azure/keyvault.go

```diff
@@ -0,0 +1,322 @@
+	"github.com/azure/spin-aks-plugin/pkg/logger"
+)
+
+type akv struct {
+	uri            string
+	id             string
+	tenantId       string
+	subscriptionId string
+	resourceGroup  string
+	name           string
+}
+
+// CertOpt specifies what kind of certificate to create
+type CertOpt func(cert *azcertificates.CreateCertificateParameters) error
+
+type Cert struct {
+	name string
+}
+
+func LoadAkv(id arm.ResourceID) *akv {
+	return &akv{
+		id:             id.String(),
+		name:           id.Name,
+		resourceGroup:  id.ResourceGroupName,
+		subscriptionId: id.SubscriptionID,
+	}
+}
+
+func ListKeyVaults(ctx context.Context, subscriptionId, resourceGroup string) ([]akv, error) {
+	lgr := logger.FromContext(ctx).With("resourceGroup", resourceGroup, "subscriptionId", subscriptionId)
+	ctx = logger.WithContext(ctx, lgr)
+	lgr.Info("starting to list keyvaults")
+	defer lgr.Info("finished listing keyvaults")
+
+	cred, err := getCred()
+	if err != nil {
+		return nil, fmt.Errorf("getting az credentials: %w", err)
+	}
+
+	client, err := armkeyvault.NewVaultsClient(subscriptionId, cred, nil)
+	if err != nil {
+		return nil, fmt.Errorf("creating client: %w", err)
+	}
+
+	pager := client.NewListByResourceGroupPager(resourceGroup, nil)
+	var akvs []akv
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("getting next page: %w", err)
+		}
+
+		for _, kv := range page.Value {
+
+			if kv == nil {
+				return nil, fmt.Errorf("nil keyvault")
+			}
+			id, err := arm.ParseResourceID(*kv.ID)
+			if err != nil {
+				return nil, fmt.Errorf("parsing resource id: %w", err)
+			}
+			newAkv := LoadAkv(*id)
+			newAkv.uri = *kv.Properties.VaultURI
+
+			akvs = append(akvs, *newAkv)
+			lgr.Info("keyvault", "name", kv.Name)
+		}
+	}
+	return akvs, nil
+}
+
+
+func NewAkv(ctx context.Context, tenantId, subscriptionId, resourceGroup, name, location string) (*akv, error) {
+	name = truncate(name, 24)
+
+	lgr := logger.FromContext(ctx).With("name", name, "resourceGroup", resourceGroup, "location", location, "subscriptionId", subscriptionId)
+	ctx = logger.WithContext(ctx, lgr)
+	lgr.Info("starting to create akv")
+	defer lgr.Info("finished creating akv")
+
+	cred, err := getCred()
+	if err != nil {
+		return nil, fmt.Errorf("getting az credentials: %w", err)
+	}
+
+	factory, err := armkeyvault.NewClientFactory(subscriptionId, cred, nil)
+	if err != nil {
+		return nil, fmt.Errorf("creating client factory: %w", err)
+	}
+
+	clientObjectId, err := getObjectId(ctx, cred)
+	if err != nil {
+		return nil, fmt.Errorf("getting client object id: %w", err)
+	}
+
+	v := &armkeyvault.VaultCreateOrUpdateParameters{
+		Location: to.Ptr(location),
+		Properties: &armkeyvault.VaultProperties{
+			AccessPolicies: []*armkeyvault.AccessPolicyEntry{
+				{
+					ObjectID: to.Ptr(clientObjectId),
+					Permissions: &armkeyvault.Permissions{
+						Certificates: []*armkeyvault.CertificatePermissions{
```

Comment:
we are only granting permissions? we should grant ability to rotate secrets to the caller maybe? 

we don't need to do anything with certs. Lots to rip out of this file

Reply:
yep